qchem/pennylane_qchem/qchem/obs.py

```diff
@@ -444,10 +444,308 @@ def particle_number(orbitals, mapping="jordan_wigner", wires=None):
+    ...                             [ 2.86964662e-16,  1.93277291e-16,  2.55224784e-16,
+    ...                              -2.77555756e-16, -1.13579985e+00,  2.06665432e-16],
+    ...                             [-3.46772026e-02,  5.27078882e-02, -3.04455743e-02,
+    ...                              -2.36158697e-16,  2.06665432e-16, -9.50966595e-01]]
+
+    >>> table, t_core = one_particle(matrix_elements, core=[0], active=[1, 2])
+    >>> print(table)
+    [[ 0.          0.         -1.49264622]
+     [ 1.          1.         -1.49264622]
+     [ 0.          2.          0.03289281]
+     [ 1.          3.          0.03289281]
+     [ 2.          0.          0.03289281]
+     [ 3.          1.          0.03289281]
+     [ 2.          2.         -1.12554473]
+     [ 3.          3.         -1.12554473]]
+    >>> print(t_core)
+    -9.45478626
+    """
+
+    orbitals = matrix_elements.shape[0]
+
+    if matrix_elements.ndim != 2:
+        raise ValueError(
+            "'matrix_elements' must be a 2D array; got matrix_elements.ndim = {}".format(
+                matrix_elements.ndim
+            )
+        )
+
+    if core is None:
+        t_core = 0
+    else:
+        if True in [i > orbitals - 1 or i < 0 for i in core]:
+            raise ValueError(
+                "Indices of core orbitals must be between 0 and {}; got core = {}".format(
+                    orbitals, core
+                )
+            )
+
+        # Compute contribution due to core orbitals
+        t_core = 2 * sum([matrix_elements[alpha, alpha] for alpha in core])
+
+    if active is None:
+        if core is None:
+            active = list(range(orbitals))
+        else:
+            active = [i for i in range(orbitals) if i not in core]
+
+    if True in [i > orbitals - 1 or i < 0 for i in active]:
+        raise ValueError(
+            "Indices of active orbitals must be between 0 and {}; got active = {}".format(
+                orbitals, active
+            )
+        )
+
+    # Indices of the matrix elements with absolute values >= cutoff
+    indices = np.nonzero(np.abs(matrix_elements) >= cutoff)
+
+    # Single out the indices of active orbitals
+    num_indices = len(indices[0])
+    pairs = [
+        [indices[0][i], indices[1][i]]
+        for i in range(num_indices)
+        if all(indices[j][i] in active for j in range(len(indices)))
+    ]
+
+    # Building the table of indices and matrix elements
+    table = np.zeros((2 * len(pairs), 3))
+
+    for i, pair in enumerate(pairs):
+        alpha, beta = pair
+        element = matrix_elements[alpha, beta]
+
+        # spin-up term
+        table[2 * i, 0] = 2 * active.index(alpha)
+        table[2 * i, 1] = 2 * active.index(beta)
+        table[2 * i, 2] = element
+
+        # spin-down term
+        table[2 * i + 1, 0] = 2 * active.index(alpha) + 1
+        table[2 * i + 1, 1] = 2 * active.index(beta) + 1
+        table[2 * i + 1, 2] = element
```

Comment:
another potential vectorization 😆 It will probably be very ugly though, so it depends on the speed up needed

Reply:
At this point we have already dropped the negligible matrix elements and selected only a few active orbitals. I don't think it is critical here and keeping it explicitly will be helpful to remember the spin index mappings.

Reply:
Sounds good, thanks @agran2018!